src/Features/LanguageServer/Protocol/RequestDispatcher.cs

```diff
@@ -9,91 +9,126 @@
                 }
             }
 
             return requestHandlerDictionary.ToImmutable();
+
+            static string GetRequestHandlerMethod(Type handlerType)
+            {
+                // Get the LSP method name from the handler's method name attribute.
+                var methodAttribute = Attribute.GetCustomAttribute(handlerType, typeof(MethodAttribute)) as MethodAttribute;
+                Contract.ThrowIfNull(methodAttribute, $"{handlerType.FullName} is missing Method attribute");
+
+                return methodAttribute.Method;
+            }
         }
 
-        public Task<ResponseType?> ExecuteRequestAsync<RequestType, ResponseType>(
-            RequestExecutionQueue queue,
+        /// <summary>
+        /// Retrieves the generic argument information from the request handler type without instantiating it.
+        /// </summary>
+        private static (Type requestType, Type responseType) ConvertHandlerTypeToRequestResponseTypes(Type handlerType)
+        {
+            var requestHandlerGenericType = handlerType.GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,>)).SingleOrDefault();
+            Contract.ThrowIfNull(requestHandlerGenericType, $"Provided handler type {handlerType.FullName} does not implement IRequestHandler<,>");
+
+            var genericArguments = requestHandlerGenericType.GetGenericArguments();
+            Contract.ThrowIfFalse(genericArguments.Length == 2, $"Provided handler type {handlerType.FullName} does not have exactly two generic arguments");
+            var requestType = genericArguments[0];
+            var responseType = genericArguments[1];
+
+            return (requestType, responseType);
+        }
+
+        public async Task<TResponseType?> ExecuteRequestAsync<TRequestType, TResponseType>(
             string methodName,
-            RequestType request,
+            TRequestType request,
             LSP.ClientCapabilities clientCapabilities,
             string? clientName,
-            CancellationToken cancellationToken) where RequestType : class
+            RequestExecutionQueue queue,
+            CancellationToken cancellationToken) where TRequestType : class
         {
-            Contract.ThrowIfNull(request);
-            Contract.ThrowIfTrue(string.IsNullOrEmpty(methodName), "Invalid method name");
-
-            if (request is ExecuteCommandParams executeCommandRequest)
-            {
-                // If we have a workspace/executeCommand request, get the request name
-                // from the command name.
-                methodName = AbstractExecuteWorkspaceCommandHandler.GetRequestNameForCommandName(executeCommandRequest.Command);
-            }
+            // Get the handler matching the requested method.
+            var requestHandlerMetadata = new RequestHandlerMetadata(methodName, typeof(TRequestType), typeof(TResponseType));
```

Comment:
Do we rely on having the same method name with potentially different request/response types to be used as a key like this? I would have assumed that we'd have the method name only and instead we'd flag/throw on a mismatch?

Reply:
True this probably could be done by method name only.  For a single dispatcher (aka server) we only allow one method with one set of types.  Updated in https://github.com/dotnet/roslyn/pull/59600